### PR TITLE
Adding workflows for unit tests and linting

### LINKED ISSUE #7: Setup CI workflow to run tests, linting and vulnerability checks
## Summary
Create CI workflow that runs tests, linting and vulnerability checks on every commit.

## Acceptance Criteria
- [ ] CI step for running unit tests
- [ ] CI step for linting the source code
- [ ] CI step for running vulnerability checks

## Details / Tasks
Unit and linting should be required for a PR to be accepted, but the vulnerability checks can be informational for now.

### PR DESCRIPTION
**Pull request by ericbuckley**
## Description
Adding 4 different GitHub workflows to help with managing code checks.
1. unit tests
2. linting
3. code vulnerabilities
4. dependency vulnerabilities

## Related Issues
closes #7 

## Additional Notes
Also removing the `scripts/vulnerability_checks.sh` script as we're going to try using GH tools (ie CodeQL and Dependabot) for managing those checks and won't need an independent script for it anymore.

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [x] I have added or updated test cases to cover my changes, if applicable.
- [x] I have minimized the number of reviewers to include only those essential for the review.
- [x] I have notified teammates in the review thread to build awareness.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

File: .github/dependabot.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+version: 2
+updates:
+  - package-ecosystem: "pip"
+    directory: "/"
+    schedule:
+      interval: "weekly"
+    commit-message:
+      prefix: "[deps] "
+    open-pull-requests-limit: 3
```
Comment (alhayward) [inline .github/dependabot.yml:9]: Does this mean we can only have 3 open PRs at a time for good practice?

File: .github/workflows/check_lint.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+name: lint check
+
+# When the workflow will be triggered
```
Comment (alhayward) [inline .github/workflows/check_lint.yml:3]: _nit:_ I'm in favor of consistent commenting. We could either add this comment and the following (`# Step 1: Check out the repository` etc.) to `.github/workflows/check_code_vulnerabilities.yml`, or delete these comments altogether from `.github/workflows/check_lint.yml` and `.github/workflows/check_unit_tests.yml`, if we think the code reads for itself (for example, the `name` arg seems to be communicating the same information).

File: .github/workflows/check_unit_tests.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        run: |
+          until pg_isready -h localhost -p 5432 -U postgres; do
+            echo "Waiting for PostgreSQL to be ready..."
+            sleep 1
+          done
+
+      # Step 5: Run the tests
+      - name: Run unit tests
+        env:
+          MPI_DB_TYPE: postgres
```
Comment (alhayward) [inline .github/workflows/check_unit_tests.yml:58]: Will these env vars ultimately be parameterized?

**Review by alhayward — APPROVED** [review]
I'm less familiar with GH Workflows, CodeQL, and Dependabot, so I defer to @derekadombek, but LGMT!

File: .github/dependabot.yml (hunk 1 above)
Comment (ericbuckley) [reply .github/dependabot.yml:9]: I believe this means that dependabot will at max have 3 open PR for dependency updates.

File: .github/workflows/check_unit_tests.yml (hunk 3 above)
Comment (ericbuckley) [reply .github/workflows/check_unit_tests.yml:58]: Eventually we'll switch to using 1 DB_URI rather value than 6 different variables to specify the database.  Is this what you mean?

File: .github/workflows/check_lint.yml (hunk 2 above)
Comment (ericbuckley) [reply .github/workflows/check_lint.yml:3]: I'll clean them up a bit and only leave in comments where I think its useful.  FWIW, chatgpt auto generated some of these initial workflows, hence the duplicative nature between names and comments.  I should have taken a deeper look before committing.

File: .github/workflows/check_unit_tests.yml [hunk 4]
```diff
@@ -13,50 +12,35 @@ jobs:
...
-            sleep 1
-          done
+          until pg_isready -U postgres; do sleep 1; done
 
-      # Step 5: Run the tests
       - name: Run unit tests
         env:
           MPI_DB_TYPE: postgres
-          MPI_DBNAME: testdb
+          MPI_DBNAME: postgres
```
Comment (alhayward) [inline .github/workflows/check_unit_tests.yml:43]: Will there be any issues with `MPI_DBNAME` and `MPI_USER` env vars having the same value?

File: .github/workflows/check_unit_tests.yml [hunk 5]
```diff
@@ -13,50 +12,35 @@ jobs:
...
       postgres:
-        image: postgres:13  # Use the latest version of PostgreSQL you need
+        image: postgres:13
         ports:
           - 5432:5432
         env:
-          POSTGRES_USER: postgres
           POSTGRES_PASSWORD: pw
-          POSTGRES_DB: testdb
-        options: >-
```
Comment (alhayward) [inline .github/workflows/check_unit_tests.yml]: For my own learning, what tests were these that were removed and why?

**Review by alhayward — DISMISSED** [review]

File: .github/workflows/check_unit_tests.yml (hunk 4 above)
Comment (ericbuckley) [reply .github/workflows/check_unit_tests.yml:43]: Nope, in fact thats the default https://hub.docker.com/_/postgres.

> POSTGRES_DB
> 
> This optional environment variable can be used to define a different name for the default database that is created > when the image is first started. If it is not specified, then the value of POSTGRES_USER will be used.

File: .github/workflows/check_unit_tests.yml (hunk 5 above)
Comment (ericbuckley) [reply .github/workflows/check_unit_tests.yml]: It seemed redundant.  We shouldn't have to tell the service to wait AND tell the tests to wait (lines 36-38) for postgres to be available.  Having the latter seemed sufficient.

File: .github/workflows/check_unit_tests.yml (hunk 4 above)
Comment (ericbuckley) [reply .github/workflows/check_unit_tests.yml]: I'm going to revert this change for now.  Many, but not all of the tests are hard coding these values and not reading them from the environment.    I'll make some more comprehensive changes in the future, but want to do that in a differently scoped PR.

File: .github/workflows/check_unit_tests.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
+name: unit tests check
+
+on:
+  push:
+    branches: [main]
```
Comment (derekadombek) [inline .github/workflows/check_unit_tests.yml:5]: Are we certain we want to unit test only when pushing to main or should we be testing for every commit upstream? Also, im guessing these were all tested on your feature branch?

**Review by alhayward — APPROVED** [review]
LGTM. Thanks for your responses!

**Pull request merged**